File: snake-strategy-bono/src/main/java/co/electric/snake/strategy/bonostrategy/directionprocessor/EquivalentBestDirectionsProcessor.java
package co.electric.snake.strategy.bonostrategy.directionprocessor;

import co.electric.snake.framework.model.Arena;
import co.electric.snake.framework.model.Coordinate;
import co.electric.snake.framework.model.Snake;
import co.electric.snake.strategy.bonostrategy.SimpleDirectionContainer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;
import java.util.Optional;
import java.util.TreeMap;

public class EquivalentBestDirectionsProcessor {

    private static final Logger LOG = LoggerFactory.getLogger(EquivalentBestDirectionsProcessor.class);

    public SimpleDirectionContainer getDirections(Snake snake, Arena arena, SimpleDirectionContainer filteredDirections) {
        final Coordinate headCoordinate = snake.getHeadCoordinate();
        final Map<Integer, SimpleDirectionContainer> distancesToFood = getDistancesToFood(arena, headCoordinate, filteredDirections);
        final SimpleDirectionContainer equivalentBestDirections = distancesToFood.values().stream()
                .findFirst()
                .orElse(new SimpleDirectionContainer());
        LOG.info("Equivalent Best Directions: " + equivalentBestDirections);
        return equivalentBestDirections;
    }

    private Map<Integer, SimpleDirectionContainer> getDistancesToFood(Arena arena, Coordinate headCoordinate, SimpleDirectionContainer filteredDirections) {
        final Map<Integer, SimpleDirectionContainer> distancesToFood = new TreeMap<>();
        final Coordinate foodCoordinate = arena.getFoodInNewList().get(0).getCoordinate();
        final Coordinate maxCoordinate = arena.getMaxCoordinate();
        filteredDirections.forEach(
                direction -> {
                    final Coordinate nextCoordinate = arena.nextCoordinate(headCoordinate, direction);
                    final int distanceToFood = nextCoordinate.minDistance(foodCoordinate, maxCoordinate);
                    final SimpleDirectionContainer directions = Optional.ofNullable(distancesToFood.get(distanceToFood)).orElse(new SimpleDirectionContainer());
                    directions.add(direction);
                    distancesToFood.put(distanceToFood, directions);
                }
        );
        LOG.info("Distances To Food: " + distancesToFood);
        return distancesToFood;
    }

}

File: snake-strategy-bono/src/main/kotlin/co/electric/snake/strategy/bonostrategy/directionprocessor/EquivalentBestDirectionsProcessor.kt
package co.electric.snake.strategy.bonostrategy.directionprocessor

import co.electric.snake.framework.model.Arena
import co.electric.snake.framework.model.Coordinate
import co.electric.snake.framework.model.Snake
import co.electric.snake.strategy.bonostrategy.SimpleDirectionContainer
import org.slf4j.LoggerFactory
import java.util.*

class EquivalentBestDirectionsProcessor {

    companion object {
        private val LOG = LoggerFactory.getLogger(EquivalentBestDirectionsProcessor::class.java)
    }

    fun getDirections(snake: Snake, arena: Arena, filteredDirections: SimpleDirectionContainer): SimpleDirectionContainer {
        val headCoordinate = snake.getHeadCoordinate()
        val distancesToFood = getDistancesToFood(arena, headCoordinate, filteredDirections)
        val equivalentBestDirections = distancesToFood.values.stream()
                .findFirst()
                .orElse(SimpleDirectionContainer())
        LOG.info("Equivalent Best Directions: $equivalentBestDirections")
        return equivalentBestDirections
    }

    private fun getDistancesToFood(arena: Arena, headCoordinate: Coordinate, filteredDirections: SimpleDirectionContainer): Map<Int, SimpleDirectionContainer> {
        val distancesToFood: MutableMap<Int, SimpleDirectionContainer> = TreeMap()
        val foodCoordinate = arena.getFoodInNewList()[0].coordinate
        val maxCoordinate = arena.maxCoordinate
        filteredDirections.forEach {
            val nextCoordinate = arena.nextCoordinate(headCoordinate, it)
            val distanceToFood = nextCoordinate.minDistance(foodCoordinate, maxCoordinate)
            val directions = Optional.ofNullable(distancesToFood[distanceToFood]).orElse(SimpleDirectionContainer())
            directions.add(it)
            distancesToFood[distanceToFood] = directions
        }
        LOG.info("Distances To Food: $distancesToFood")
        return distancesToFood
    }

}
